pkg/controller/node/node_controller.go

```diff
@@ -128,6 +143,199 @@ func (ctrl *Controller) Run(workers int, stopCh <-chan struct{}) {
+	switch {
+	case desiredNumberOfMastersSchedulable == len(currentSchedulableMasters):
+		glog.V(4).Infof("Desired number of masters are already schedulable")
+
+	case desiredNumberOfMastersSchedulable > len(currentSchedulableMasters):
+		neededSchedulableMasters := desiredNumberOfMastersSchedulable - len(currentSchedulableMasters)
+		crtl.makeMasterNodesSchedulable(currentUnSchedulableMasters, neededSchedulableMasters)
+
+	case desiredNumberOfMastersSchedulable < len(currentSchedulableMasters):
+		neededUnSchedulableMasters := desiredNumberOfMastersSchedulable - len(currentSchedulableMasters)
+		crtl.makeMasterNodesUnSchedulable(currentSchedulableMasters, neededUnSchedulableMasters)
+	}
+}
+
+// getSchedulableAndUnSchedulableMasterNodes takes the currentMasters and splits it into 2 slices of
+// masterNodes which are schedulable and unschedulable respectfully.
+func getSchedulableAndUnSchedulabeMasterNodes(currentMasters []*corev1.Node) ([]*corev1.Node, []*corev1.Node) {
+	var currentSchedulableMasters = make([]*corev1.Node, 0, len(currentMasters))
+	var currentUnSchedulableMasters = make([]*corev1.Node, 0, len(currentMasters))
+	for _, node := range currentMasters {
+		_, hasWorkerLabel := node.Labels["node-role.kubernetes.io/worker"]
+		hasMasterTaint := false
+		for _, taint := range node.Spec.Taints {
+			if taint.Key == "node-role.kubernetes.io/master" && taint.Effect == corev1.TaintEffectNoSchedule {
+				hasMasterTaint = true
+			}
+		}
+		if hasWorkerLabel && !hasMasterTaint {
+			currentSchedulableMasters = append(currentSchedulableMasters, node)
+		} else {
+			currentUnSchedulableMasters = append(currentUnSchedulableMasters, node)
+		}
+	}
+	return currentSchedulableMasters, currentUnSchedulableMasters
+}
+
+// This is a safe guard to ensure that we're having acceptable values(We'll have one more validation on the api side).
+// The acceptable values currently are 100% or 0%. In future, we will relax this restriction.
+// TODO: @ravig remove this validation when we relax this restriction and this code could be in api validation when
+// it doesn't depend on number of master nodes
+func validateDesiredSchedulableMasterField(desiredSchedulableMasters intstrutil.IntOrString, currentMasterNodesCount int) {
+	switch desiredSchedulableMasters.Type {
+	case intstrutil.String:
+		maxMastersToBeSchedulable := "100%"
+		minMastersToBeSchedulable := "0%"
+		if desiredSchedulableMasters != intstrutil.FromString(maxMastersToBeSchedulable) || desiredSchedulableMasters != intstrutil.FromString(minMastersToBeSchedulable) {
+			glog.Errorf("Acceptable values for percentangeOfMastersSchedulable are 100 and 0 but found %v", desiredSchedulableMasters)
+			return
+		}
+	case intstrutil.Int:
+		if desiredSchedulableMasters.IntValue() == 0 {
+			desiredSchedulableMasters = intstrutil.FromInt(0)
+		}
+		// If this value is greater than no of master nodes, so set it to all the master available.
```

Comment:
Comment is a lie...this is basically always assigning it to `currentMasterNodesCount` right?